src/target/stm32f1_ch32.c

```diff
@@ -0,0 +1,338 @@
+		return -1;
+	}
+	return 0;
+}
+
+/**
+	\fn waitFlashReady
+	\brief poll the beginning of a block till it is fffff, meaning we can proceeed
+
+*/
+
+static bool waitFlashReady(target *t,uint32_t adr)
+{
+  // Wait a bit
+  // the # of cycles needed to do N read access over SWD
+  // the actual number is around 10
+  // then check we do read ffff (/!\ even if we read ffff it does not mean its ok
+  // these are the data from the previous operation and they could be ffff)
+  uint32_t ff;
+  for(int i=0;i<32;i++)   ff=target_mem_read32(t,adr);
+  if(ff!=0xffffffffUL) {
+	  ERROR_CH("ch32f1 Not erased properly at %x or flash access issue\n",adr);
+	  return false;
+  }
+  return true;
+}
+/**
+  \fn ch32f1_flash_write
+  \brief fast flash for ch32. Load 128 bytes chunk and then flash them
+*/
+
+static int upload(target *t, uint32_t dest, const void  *src, uint32_t offset)
+{
+	const uint32_t *ss=(const uint32_t *)(src+offset);
+	uint32_t dd=dest+offset;
+
+	SET_CR(FLASH_CR_FTPG_CH32);
+	target_mem_write32(t, dd+0,ss[0]);
+	target_mem_write32(t, dd+4,ss[1]);
+	target_mem_write32(t, dd+8,ss[2]);
+	target_mem_write32(t, dd+12,ss[3]);
+	SET_CR(FLASH_CR_BUF_LOAD_CH32); /* BUF LOAD */
+	WAIT_EOP();
+	CLEAR_EOP();
+	CLEAR_CR(FLASH_CR_FTPG_CH32);
+	MAGIC((dest+offset));
+	return 0;
+}
+/**
+	\fn ch32_buffer_clear
+	\brief clear the write buffer
+*/
+int ch32_buffer_clear(target *t)
+{
+  SET_CR(FLASH_CR_FTPG_CH32); // Fast page program 4-
+  SET_CR(FLASH_CR_BUF_RESET_CH32); // BUF_RESET 5-
+  WAIT_BUSY(); // 6-
+  CLEAR_CR(FLASH_CR_FTPG_CH32); // Fast page program 4-
+  return 0;
+}
+#define CH32_VERIFY
+
+/**
+
+*/
+static int ch32f1_flash_write(struct target_flash *f,
+	                           target_addr dest, const void *src, size_t len)
+{
+	target *t = f->t;
+	size_t length = len;
+#ifdef CH32_VERIFY
+	target_addr orgDest=dest;
+	const void *orgSrc=src;
+#endif
+	DEBUG_CH("CH32: flash write 0x%x ,size=%d\n",dest,len);
+
+	while(length>0)
+	{
+		if(ch32f1_flash_unlock(t)) {
+			ERROR_CH("ch32f1 cannot fast unlock\n");
+			return -1;
+		}
+		WAIT_BUSY();
+
+		// Buffer reset...
+		ch32_buffer_clear(t);
+		// Load 128 bytes to buffer
+		if(!waitFlashReady(t,dest)) {
+			return -1;
+		}
+		for(int i=0;i<8;i++) {
+			if(upload(t,dest,src, 16*i)) {
+	  			ERROR_CH("Cannot upload to buffer\n");
+				return -1;
+			}
+		}
+		// write buffer
+		SET_CR(FLASH_CR_FTPG_CH32);
+		target_mem_write32(t, FLASH_AR, dest); // 10
+		SET_CR(FLASH_CR_STRT); // 11 Start
+		WAIT_EOP(); // 12
+		CLEAR_EOP();
+		CLEAR_CR(FLASH_CR_FTPG_CH32);
+
+		MAGIC((dest));
+
+		// next
+		if(length>128)
+		  length-=128;
+		else
+		  length=0;
+		dest+=128;
+		src+=128;
+
+		sr = target_mem_read32(t, FLASH_SR); // 13
+		ch32f1_flash_lock(t);
+		if ((sr & SR_ERROR_MASK) ) {
+			ERROR_CH("ch32f1 flash write error 0x%" PRIx32 "\n", sr);
+			return -1;
+		}
+
+	}
+#ifdef CH32_VERIFY
+	DEBUG_CH("Verifying\n");
+	int i=0;
+	for(i=0;i<(int)len;i+=4)
```

Comment:
use size_t for the loop counter, don't cast a size_t variable to int. The compiler warning was telling you about a bug you've created here - this is not the fix.